model/vfs/vfsswift/impl_v3.go

```diff
@@ -259,6 +249,65 @@ func (sfs *swiftVFSV3) CreateFile(newdoc, olddoc *vfs.FileDoc, opts ...vfs.Creat
 	}, nil
 }
 
+func (sfs *swiftVFSV3) CopyFile(olddoc, newdoc *vfs.FileDoc) error {
+	if lockerr := sfs.mu.Lock(); lockerr != nil {
+		return lockerr
+	}
+	defer sfs.mu.Unlock()
+
+	newsize, maxsize, capsize, err := vfs.CheckAvailableDiskSpace(sfs, newdoc)
+	if err != nil {
+		return err
+	}
+	if newsize > maxsize {
+		return vfs.ErrFileTooBig
+	}
+
+	newname := newdoc.DocName
+	exists, err := sfs.Indexer.DirChildExists(newdoc.DirID, newname)
+	if err != nil {
+		return err
+	}
+	if exists {
+		newname = vfs.ConflictName(sfs, newdoc.DirID, newname, true)
+		exists, err = sfs.Indexer.DirChildExists(newdoc.DirID, newname)
+		if err != nil {
+			return err
+		}
+		if exists {
+			return os.ErrExist
+		}
+	}
+	newdoc.DocName = newname
```

Comment:
I'm not sure if it is needed, but I would have added:

```go
newdoc.ResetFullPath()
```

The `file` has a `fullpath` field, which is not saved in CouchDB but computed from the path of its parent, and kept as cache. When we change the name or the parent, we often reset the `fullpath` to avoid keeping a cached value with an incorrect path.

Reply:
It seems necessary indeed. I call the method after cloning the old document but I didn't think about calling it again here. I'll fix that.